mm/ktsan/sync_atomic.c

```diff
@@ -111,21 +105,17 @@ static kt_tab_sync_t *kt_atomic_post_op(kt_thr_t *thr, uptr_t pc, uptr_t addr,
 		thr->acquire_active = KT_TAME_COUNTER_LIMIT;
 	}
 
-	/* This will catch races between atomic operations and non-atomic
-	 * writes (in particular with kfree).
-	 */
-	if (read && !write)
-		kt_access(thr, pc, addr, KT_ACCESS_SIZE_1, true);
-
 	return sync;
 }
 
-#define KT_ATOMIC_OP(op, ad, mo, read, write)				\
+#define KT_ATOMIC_OP(op, ad, mo, size, read, write)			\
 do {									\
 	kt_tab_sync_t *sync = NULL;					\
 									\
 	sync = kt_atomic_pre_op(thr, pc, ad, mo, read, write, sync);	\
 									\
+	kt_access(thr, pc, ad, size, !write, true);			\
```

Comment:
Consider also patterns when you read/write/allocate/free the variable before/after the atomic access.